Login with Facebook email

alexanderpetkov opened the pull request.

Commit:
Add initial facebook login

Commit:
Handle Facebook login failure. Refactor sessions_controller.rb

Commit:
Test successful Facebook login

Commit:
Expand facebook login tests

Commit:
Style Facebook button

valo commented on app/controllers/sessions_controller.rb:
```diff
@@ -35,4 +19,56 @@ def destroy
+    omniauth_response = request.env['omniauth.auth']
+    facebook_email = omniauth_response['info']['email'] if omniauth_response.is_a?(Hash)
+
+    user = User.where(email: facebook_email).last
```
Can we extend the login to also create a new user account if one doesn't exist yet?

valo commented on db/migrate/20160205194153_add_omniauth_to_users.rb:
```diff
@@ -0,0 +1,6 @@
+class AddOmniauthToUsers < ActiveRecord::Migration
+  def change
+    add_column :users, :provider, :string
+    add_column :users, :uid, :string
```
These columns are not used at the moment, but if account creation is implemented I think they will be useful ;-)

valo commented on app/controllers/sessions_controller.rb:
```diff
@@ -10,24 +10,8 @@ def new
-      @login       = params[:login]
-      @remember_me = params[:remember_me]
-      render :action => 'new'
-    end
```
I really like how you broke down this function into smaller pieces! Good job!

valo commented on spec/controllers/sessions_controller_spec.rb:
```diff
@@ -0,0 +1,45 @@
+  #     it "redirects to root" do
+  #       expect(response).to redirect_to root_path
+  #     end
+  # end
```
This is annoying, but it seems there is no easy solution for it. My guess is that this route is handled by a rack middleware and this is the reason we can't invoke it from the specs.

Commit:
Enable creating facebook users on DB level

Commit:
Create user when logging with new facebook email

Commit:
Write specs for registering with facebook

Commit:
Clean out facebook vs password login logic:

When having only standard account, FB login logs you with it;
When having only FB account, password login prompts you to log with FB;
When having both accounts, log the standard account via password and the FB account via FB.
Note: you can have both accounts if you register with FB and later register the same mail via the login form.

Commit:
Write tests for complex facebook vs password login logic

Commit:
Make User.maycamp method cover users with null provider as well

valo commented on app/models/user.rb:
```diff
@@ -56,6 +68,39 @@ def self.authenticate(login, password)
+      provider: providers[provider],
+      provider_id: provider_id.to_s
+    )
+  end
```
I think it will be nice to have specs for these new class methods

valo commented on app/models/user.rb:
```diff
@@ -56,6 +68,39 @@ def self.authenticate(login, password)
+      email: email,
+      provider: providers[provider],
+      provider_id: provider_id.to_s
+    )
```
I am wondering if there is a way to set the `city` here? It it used in several places and I am not sure if the app will work with accounts without `city`. Have you tried if everything renders properly without city?

valo commented on app/models/user.rb:
```diff
@@ -47,7 +58,8 @@ class User < ActiveRecord::Base
   def self.authenticate(login, password)
-    user = find_by_login(login.downcase) || find_by_email(login.downcase) # need to get the salt
+    user = find_by_login(login.downcase) ||
+           User.maycamp.find_by_email(login.downcase) # need to get the salt
```
This new `maycamp` scope is nice. I am wondering if it is a good idea to have a spec for it...

Commit:
Request user_location Facebook permission

Commit:
Store user city on FB registration

Commit:
Write specs for newly created User methods
